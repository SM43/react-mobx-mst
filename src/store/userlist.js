import { types, getEnv, flow } from "mobx-state-tree";
import { values } from "mobx";

export const User = types.model({
  id: types.integer,
  name: types.optional(types.string, ""),
  age: types.optional(types.integer, 0),
  class: types.optional(types.string, ""),
});

export const Store = types
  .model({
    users: types.array(User),
    filter: types.optional(types.array(types.string), []),
    sort: types.optional(types.string, ""),
    isLoading: true,
  })
  .views((self) => ({
    get count() {
      return self.users.length;
    },
    get userList() {
      return filterAndSortUsers(values(self.users), self.filter, self.sort);
    },
    get fetch() {
      return getEnv(self).fetch;
    },
  }))
  .actions((self) => ({
    afterCreate() {
      self.users.clear();
      loadUsers(self);
    },
    add(item) {
      self.users.push(item);
    },
    setSort(sort) {
      self.sort = sort;
    },
    setFilter(filter) {
      self.filter = filter;
    },
    setStatus(status) {
      self.isLoading = status;
    },
  }));

const loadUsers = flow(function* loadUsers(self) {
  try {
    const json = yield self.fetch("./users.json");
    json.forEach((userJson) => {
      self.add(userJson);
    });
    self.setStatus(false)
  } catch (err) {
    console.error("Failed to load users ", err);
  }
});

function filterAndSortUsers(users, filter, sort) {
  if (filter.length === 0 && sort === "") {
    return users;
  }
  if (filter.length === 0) {
    if (sort === "name") {
      return users.sort((a, b) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
    } else if (sort === "age") {
      return users.sort((a, b) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
    } else {
      return users;
    }
  }
  if (sort === "name") {
    return users
      .filter((u) => filter.includes(u.class))
      .sort((a, b) => (a.name > b.name ? 1 : a.name === b.name ? 0 : -1));
  } else if (sort === "age") {
    return users
      .filter((u) => filter.includes(u.class))
      .sort((a, b) => (a.age > b.age ? 1 : a.age === b.age ? 0 : -1));
  } else {
    return users.filter((u) => filter.includes(u.class));
  }
}
