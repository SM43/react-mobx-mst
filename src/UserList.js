import React from "react";
import { useObserver } from "mobx-react";
import { StoreContext } from "./store/store";
// import {
//   Table,
//   TableHeader,
//   TableBody,
//   sortable,
// } from "@patternfly/react-table";

export const UserList = () => {
  const store = React.useContext(StoreContext);
  var userArr;
  // var rows = [];
  // var columns = [
  //   { title: "Name", transforms: [sortable] },
  //   { title: "Age", transforms: [sortable] },
  //   { title: "Class" },
  // ];
  return useObserver(() => (
    <div>
      <h2>No. of Users: {store.userList.length} </h2>
      <ul>
        {
          ((userArr = store.userList),
          userArr.map((user) => (
            <li key={user.name}>
              Name: {user.name} &emsp; Age: {user.age} &emsp; Class:{" "}
              {user.class}
            </li>
          )))
        }
      </ul>

      {/* {
        ((userArr = store.userList),
        userArr.map((user) =>
          rows.push({ cells: [user.name, user.age, user.class] })
        ))
      }
      <Table
        aria-label="Sortable Table"
        caption="Users"
        cells={columns}
        rows={rows}
      >
        <TableHeader />
        <TableBody />
      </Table> */}
    </div>
  ));
};
