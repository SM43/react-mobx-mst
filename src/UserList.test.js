import React from "react";
import renderer from "react-test-renderer";
import fs from "fs";
import { Store } from "./store/userlist";
import { StoreContext } from "./store/store";
import "./App.css";
import { UserFilter, UserList } from "./UserList.js";
import { mount } from "enzyme";

const userFetcher = () =>
  Promise.resolve(JSON.parse(fs.readFileSync("./src/store/users.json")));

const store = Store.create({}, { fetch: userFetcher });

const StoreProvider = ({ children }) => {
  return (
    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  );
};

describe("<UserList />", () => {
  it("should render correctly", () => {
    const component = renderer.create(<StoreProvider><UserList /></StoreProvider>);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});