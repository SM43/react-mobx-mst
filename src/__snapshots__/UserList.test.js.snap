// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> should render correctly 1`] = `
<div>
  <h2>
    No. of Users: 
    6
     
  </h2>
  <ul>
    <li>
      Name: 
      rachel
         Age: 
      18
         Class:
       
      FY
    </li>
    <li>
      Name: 
      monica
         Age: 
      21
         Class:
       
      SY
    </li>
    <li>
      Name: 
      joey
         Age: 
      27
         Class:
       
      FY
    </li>
    <li>
      Name: 
      phoebe
         Age: 
      22
         Class:
       
      TY
    </li>
    <li>
      Name: 
      ross
         Age: 
      17
         Class:
       
      FY
    </li>
    <li>
      Name: 
      chandler
         Age: 
      28
         Class:
       
      SY
    </li>
  </ul>
</div>
`;
